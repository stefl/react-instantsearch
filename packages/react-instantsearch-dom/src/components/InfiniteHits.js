import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import translatable from '../core/translatable';
import createClassNames from '../core/createClassNames';

const cx = createClassNames('InfiniteHits');

class InfiniteHits extends Component {
  render() {
    const {
      hitComponent: HitComponent,
      hits,
      hasMore,
      refine,
      translate,
      className,
    } = this.props;

    return (
      <div className={classNames(cx(''), className)}>
        <ul className={cx('list')}>
          {hits.map(hit => (
            <li key={hit.objectID} className={cx('item')}>
              <HitComponent hit={hit} />
            </li>
          ))}
        </ul>
        <button
          className={cx('loadMore', !hasMore && 'loadMore--disabled')}
          onClick={() => refine()}
          disabled={!hasMore}
        >
          {translate('loadMore')}
        </button>
      </div>
    );
  }
}

InfiniteHits.propTypes = {
  hits: PropTypes.arrayOf(PropTypes.object).isRequired,
  hasMore: PropTypes.bool.isRequired,
  refine: PropTypes.func.isRequired,
  translate: PropTypes.func.isRequired,
  className: PropTypes.string,
  hitComponent: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
};

InfiniteHits.defaultProps = {
  className: '',
  hitComponent: hit => (
    <div
      style={{
        borderBottom: '1px solid #bbb',
        paddingBottom: '5px',
        marginBottom: '5px',
        wordBreak: 'break-all',
      }}
    >
      {JSON.stringify(hit).slice(0, 100)}...
    </div>
  ),
};

export default translatable({
  loadMore: 'Load more',
})(InfiniteHits);
