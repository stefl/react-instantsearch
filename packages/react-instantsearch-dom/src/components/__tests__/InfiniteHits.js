import React from 'react';
import PropTypes from 'prop-types';
import renderer from 'react-test-renderer';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import InfiniteHits from '../InfiniteHits';

Enzyme.configure({ adapter: new Adapter() });

describe('InfiniteHits', () => {
  const Hit = ({ hit }) => <div>{JSON.stringify(hit)}</div>;

  Hit.propTypes = {
    hit: PropTypes.object,
  };

  it('accepts a hitComponent prop', () => {
    const hits = [{ objectID: 0 }, { objectID: 1 }, { objectID: 2 }];
    const tree = renderer.create(
      <InfiniteHits
        hitComponent={Hit}
        hits={hits}
        hasMore={false}
        refine={() => undefined}
      />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('accepts a custom className', () => {
    const hits = [{ objectID: 0 }, { objectID: 1 }, { objectID: 2 }];
    const tree = renderer.create(
      <InfiniteHits
        className="MyCustomInfiniteHits"
        hitComponent={Hit}
        hits={hits}
        hasMore={false}
        refine={() => undefined}
      />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('calls refine when the load more button is clicked', () => {
    const mockedRefine = jest.fn();
    const hits = [{ objectID: 0 }, { objectID: 1 }, { objectID: 2 }];
    const wrapped = mount(
      <InfiniteHits
        refine={mockedRefine}
        hitComponent={Hit}
        hits={hits}
        hasMore={true}
      />
    );
    expect(mockedRefine.mock.calls).toHaveLength(0);
    wrapped.find('button').simulate('click');
    expect(mockedRefine.mock.calls).toHaveLength(1);
  });

  it('Button is disabled when it is the last page', () => {
    const hits = [{ objectID: 0 }, { objectID: 1 }, { objectID: 2 }];
    const wrapped = mount(
      <InfiniteHits
        refine={() => undefined}
        hitComponent={Hit}
        hits={hits}
        hasMore={false}
      />
    );
    expect(wrapped.find('button').props().disabled).toBe(true);
  });
});
