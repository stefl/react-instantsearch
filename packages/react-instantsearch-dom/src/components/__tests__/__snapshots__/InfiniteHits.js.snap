// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfiniteHits accepts a custom className 1`] = `
<div
  className="ais-InfiniteHits MyCustomInfiniteHits"
>
  <ul
    className="ais-InfiniteHits-list"
  >
    <li
      className="ais-InfiniteHits-item"
    >
      <div>
        {"objectID":0}
      </div>
    </li>
    <li
      className="ais-InfiniteHits-item"
    >
      <div>
        {"objectID":1}
      </div>
    </li>
    <li
      className="ais-InfiniteHits-item"
    >
      <div>
        {"objectID":2}
      </div>
    </li>
  </ul>
  <button
    className="ais-InfiniteHits-loadMore ais-InfiniteHits-loadMore--disabled"
    disabled={true}
    onClick={[Function]}
  >
    Load more
  </button>
</div>
`;

exports[`InfiniteHits accepts a hitComponent prop 1`] = `
<div
  className="ais-InfiniteHits"
>
  <ul
    className="ais-InfiniteHits-list"
  >
    <li
      className="ais-InfiniteHits-item"
    >
      <div>
        {"objectID":0}
      </div>
    </li>
    <li
      className="ais-InfiniteHits-item"
    >
      <div>
        {"objectID":1}
      </div>
    </li>
    <li
      className="ais-InfiniteHits-item"
    >
      <div>
        {"objectID":2}
      </div>
    </li>
  </ul>
  <button
    className="ais-InfiniteHits-loadMore ais-InfiniteHits-loadMore--disabled"
    disabled={true}
    onClick={[Function]}
  >
    Load more
  </button>
</div>
`;
